RecoTracker/ExoRegionalTracking/plugins/ExoRegionProducer.cc

```diff
@@ -0,0 +1,231 @@
+    distances.sort(comp);
+    DistanceItr d = distances.begin();
+    if (d->distance() > rParam_)
+      break;
+
+    d->entities().first->merge(*d->entities().second);
+    d->entities().second->setInvalid();
+
+    const auto distancePred = [](const Distance &a) {
+      return (!a.entities().first->valid() || !a.entities().second->valid());
+    };
+    const auto pseudoROIPred = [](const ExoRegion &a) { return !a.valid(); };
+    distances.remove_if(distancePred);
+    pseudoROIs.remove_if(pseudoROIPred);
+  }
+
+  // Remove invalid ROIs.
+  const auto roiPred = [&](const ExoRegion &roi) {
+    if (!roi.valid())
+      return true;
+    const math::XYZVector x(roi.vx(), roi.vy(), roi.vz());
+    if ((minRadius_ >= 0.0 && (x - bs).rho() < minRadius_) ||
+        (genMatch_ != "None" && !roiContainsDecayVertex(x, decayVertices, roi.rParam())))
+      return true;
+    const double discriminatorValue = ((discriminatorCut_ > 0.0) ? getDiscriminatorValue(roi, *beamSpot) : 1.0);
+    if (discriminatorValue < discriminatorCut_)
+      return true;
+    return false;
+  };
+  pseudoROIs.remove_if(roiPred);
+
+  auto regionsOfInterest = make_unique<vector<reco::Vertex> >();
+  regionsOfInterest->clear();
+
+  const vector<double> error = {1.0, 0.0, 1.0, 0.0, 0.0, 1.0};
+
+  for (const auto &roi : pseudoROIs)
+    regionsOfInterest->emplace_back(reco::Vertex::Point(roi.vx(), roi.vy(), roi.vz()),
+                                    reco::Vertex::Error(error.begin(), error.end(), true, true));
+
+  event.put(move(regionsOfInterest));
+}
+
+void ExoRegionProducer::getDecayVertices(const edm::View<reco::GenParticle> &genParticles,
+                                         vector<math::XYZVector> &decayVertices) const {
+  for (const auto &genParticle : genParticles) {
+    if (genMatch_ == "HToSSTobbbb") {
+      if (genParticle.pdgId() != 5)
+        continue;
+      if (!genParticle.isHardProcess())
+        continue;
+      if (abs(genParticle.mother()->pdgId()) != 9000006)
+        continue;
+    } else if (genMatch_ == "stopToBottom") {
+      if (abs(genParticle.pdgId()) != 5)
+        continue;
+      if (!genParticle.isHardProcess())
+        continue;
+      if (abs(genParticle.mother()->pdgId()) < 1000000)
+        continue;
+    }
+
+    decayVertices.emplace_back(genParticle.vx(), genParticle.vy(), genParticle.vz());
+  }
+}
+
+const bool ExoRegionProducer::roiContainsDecayVertex(const math::XYZVector &roi,
+                                                     const vector<math::XYZVector> &decayVertices,
+                                                     const double r) const {
+  for (const auto &decayVertex : decayVertices)
+    if ((roi - decayVertex).r() < r)
+      return true;
+  return false;
+}
+
+const double ExoRegionProducer::getDiscriminatorValue(const ExoRegion &roi, const reco::BeamSpot &bs) const {
+  tensorflow::Tensor vertexTensor(tensorflow::DT_FLOAT, tensorflow::TensorShape({1, 40, 23}));
```

Comment:
could you add a comment about these "magic numbers", please ?

Reply:
I added some comments and also attempted to assign the magic numbers to meaningfully named constants.